orchestrator/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,481 @@
+package kubernetes
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+
+	"github.com/rancher/longhorn-manager/crdstore"
+	"github.com/rancher/longhorn-manager/orchestrator"
+	"github.com/rancher/longhorn-manager/types"
+	"github.com/rancher/longhorn-manager/util"
+
+	apiv1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kCli "k8s.io/client-go/kubernetes"
+)
+
+const (
+	cfgDirectory = "/var/lib/rancher/longhorn/"
+	hostUUIDFile = cfgDirectory + ".physical_host_uuid"
+)
+
+var (
+	WaitDeviceTimeout = 30 //seconds
+	WaitAPITimeout    = 30 //seconds
+	WaitPodTimeout    = 2  //seconds
+	WaitPodCounter    = 3
+)
+
+type Kubernetes struct {
+	EngineImage string
+	IP          string
+	currentNode *types.NodeInfo
+	cli         *kCli.Clientset
+}
+
+type Config struct {
+	EngineImage string
+}
+
+func NewKuberOrchestrator(cfg *Config) (*Kubernetes, error) {
+	if cfg.EngineImage == "" {
+		return nil, fmt.Errorf("missing required parameter EngineImage")
+	}
+
+	config, err := crdstore.GetClientConfig("")
+	if err != nil {
+		return nil, err
+	}
+
+	kube := &Kubernetes{
+		EngineImage: cfg.EngineImage,
+	}
+
+	kube.cli, err = kCli.NewForConfig(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot connect to kubernetes")
+	}
+
+	if err = kube.updateNetwork(); err != nil {
+		return nil, errors.Wrapf(err, "fail to detect dedicated pod network")
+	}
+
+	logrus.Infof("Detected  IP is %s", kube.IP)
+
+	if err := kube.updateCurrentNode(); err != nil {
+		return nil, err
+	}
+
+	logrus.Info("Kubernetes orchestrator is ready")
+	return kube, nil
+}
+
+func (k *Kubernetes) getCurrentNodePod() (*apiv1.Pod, error) {
+	podName, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	pod, err := k.cli.CoreV1().Pods(apiv1.NamespaceDefault).Get(podName, meta_v1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return pod, nil
+}
+
+func (k *Kubernetes) updateNetwork() error {
+	pod, err := k.getCurrentNodePod()
+	if err != nil {
+		return err
+	}
+
+	if pod.Status.PodIP == "" {
+		return fmt.Errorf("Can't get node IP")
+	}
+
+	k.IP = pod.Status.PodIP
+	return nil
+}
+
+func (k *Kubernetes) updateCurrentNode() error {
+	var err error
+	node := &types.NodeInfo{
+		IP:               k.IP,
+		OrchestratorPort: types.DefaultOrchestratorPort,
+	}
+	node.Name, err = os.Hostname()
```

Comment:
This hostname will be shown as the host for the user to choose where to attach the volume. It's useless using the container's hostname. We need to get the k8s node name or the real hostname in order to make this information useful.

Reply:
ok, It will use the real node name as the identity.